client/db/benches/state_access.rs

```diff
@@ -159,152 +204,191 @@ fn generate_storage() -> (Vec<Vec<u8>>, Vec<(Vec<u8>, Vec<u8>)>) {
 		group.bench_function(desc, |b| {
-			b.iter_batched(
-				|| backend.state_at(block_hash).expect("Creates state"),
-				|state| {
-					for key in keys.iter().cycle().take(keys.len() * multiplier) {
-						let _ = state.storage(&key).expect("Doesn't fail").unwrap();
-					}
-				},
-				BatchSize::SmallInput,
-			)
+			if repeat_shared == 1 {
+				b.iter_batched(
+					|| {
+						let backend = create_backend(config, &path);
+						backend.state_at(block_hash).expect("Creates state")
+					},
+					|state| {
+						for key in keys.iter().cycle().take(keys.len() * repeat_local) {
+							match kind {
+								AccessType::Storage => {
+									let _ = state.storage(&key).expect("Doesn't fail").unwrap();
+								},
+								AccessType::StorageHash => {
+									let _ =
+										state.storage_hash(&key).expect("Doesn't fail").unwrap();
+								},
+							}
+						}
+					},
+					BatchSize::SmallInput,
+				)
+			} else {
+				b.iter_batched(
+					|| create_backend(config, &path),
+					|backend| {
+						for _ in 0..repeat_shared {
+							// This resets the local cache and merges it with the shared cache.
+							let state = backend.state_at(block_hash).expect("Creates state");
+
+							for key in keys.iter().cycle().take(keys.len() * repeat_local) {
+								match kind {
+									AccessType::Storage => {
+										let _ = state.storage(&key).expect("Doesn't fail").unwrap();
+									},
+									AccessType::StorageHash => {
+										let _ = state
+											.storage_hash(&key)
+											.expect("Doesn't fail")
+											.unwrap();
+									},
+								}
+							}
+						}
+					},
+					BatchSize::SmallInput,
+				)
+			}
 		});
 	};
 
-	bench_multiple_values(
-		BenchmarkConfig::TrieNodeCache,
-		"with trie node cache and reading each key once",
-		1,
-	);
-	bench_multiple_values(BenchmarkConfig::NoCache, "no cache and reading each key once", 1);
-
-	bench_multiple_values(
-		BenchmarkConfig::TrieNodeCache,
-		"with trie node cache and reading 4 times each key in a row",
-		4,
-	);
-	bench_multiple_values(
-		BenchmarkConfig::NoCache,
-		"no cache and reading 4 times each key in a row",
-		4,
-	);
-
-	group.finish();
-
-	let mut group = c.benchmark_group("Reading a single value");
-
-	let mut bench_single_value = |config, desc, multiplier| {
-		let backend = create_backend(config, &path);
-
-		group.bench_function(desc, |b| {
-			b.iter_batched(
-				|| backend.state_at(block_hash).expect("Creates state"),
-				|state| {
-					for key in keys.iter().take(1).cycle().take(multiplier) {
-						let _ = state.storage(&key).expect("Doesn't fail").unwrap();
-					}
-				},
-				BatchSize::SmallInput,
-			)
-		});
-	};
+	let run_benchmark_with_and_without_cache =
+		|mut group: criterion::BenchmarkGroup<criterion::measurement::WallTime>,
+		 keys: &[Vec<u8>],
+		 repeat_shared: usize,
+		 repeat_local: usize,
+		 kind: AccessType| {
+			run_benchmark(
+				&mut group,
+				BenchmarkConfig::TrieNodeCache,
+				"with trie node cache",
```

Comment:
```suggestion
				"with cache",
```

The other naming confused me, as we have one type of cache any way no need to differentiate.